File: person.js

function Person(person_data) {
    var rect = get_person_rect(person_data);
    this.width = rect.width;
    this.height = rect.height;

    this.name = person_data.name;
    this.surname = person_data.surname;
    this.id = person_data.id;
    this.couple_id = null;
    this.parent_couple_id = null;
    this.couple_person = null;
    this.parents = null;
    this.sex = person_data.sex;
    this._x = null;
    this._y = null;
    this._layer = null;

    for (var attr_name in person_data) {
        this[attr_name] = person_data[attr_name];
    }
}

File: family/person.py
import logging

from family.colors import COLOR_FEMALE_NODE, COLOR_MALE_NODE
from .node import Node

logger = logging.getLogger(__name__)


class Person(Node):
    def __init__(self, data: dict):
        assert 'id' in data
        
        self.name = None
        self.surname = None
        self.prev_surname = None
        self.id = None
        self.sex = None

        self.couple_id = None
        self.parent_couple_id = None
        self.parents = None

        for attr_name, value in data.items():
            if hasattr(self, attr_name):
                setattr(self, attr_name, value)
            else:
                logger.warning('Unexpected field {}={}'.format(attr_name, value))
        
        super().__init__(
            data['id'],
            color=COLOR_FEMALE_NODE if self.sex == 'female' else COLOR_FEMALE_NODE,
            size=3
        )


    def sex_step(self):
        return -1 if self.sex == 'male' else 1

    def steps(self):
        edges = []
        if self.parent_couple_id:
            edges.append((self.parent_couple_id, (-1, 0)))
        if self.couple_id:
            edges.append((self.couple_id, (0, self.sex_step())))
        # Order is important for better visualization
        return edges

    @property
    def label(self):
        return '{} {}'.format(self.name, self.surname)
